scanpy/tests/test_rank_genes_groups.py

```diff
@@ -139,6 +139,20 @@ def test_results_layers():
     for name in true_scores_t_test.dtype.names:
         assert not np.allclose(true_scores_t_test[name][:7], adata.uns['rank_genes_groups']['scores'][name][:7])
 
+@pytest.mark.xfail
+def test_singlets():
+    pbmc = pbmc68k_reduced()
+    pbmc.obs['louvain'] = pbmc.obs['louvain'].cat.add_categories(['11'])
+    pbmc.obs['louvain'][0] = '11' 
+    
+    rank_genes_groups(pbmc, groupby = 'louvain')
+
+@pytest.mark.xfail
+def test_emptycat():
+    pbmc = pbmc68k_reduced()
+    pbmc.obs['louvain'] = pbmc.obs['louvain'].cat.add_categories(['11'])
+    
+    rank_genes_groups(pbmc, groupby = 'louvain')
```

Comment:
I think we implicitly remove categories that have no entries. Is this not the case?

Reply:
I don't think so- we managed to break things by having an empty category https://github.com/ebi-gene-expression-group/scanpy-scripts/pull/88#issuecomment-726001228

Reply:
Though to be fair we did it by editing .obs directly, so it may not be something you need to expect.

Reply:
Ah, I see. This  looks like a bug to me, since we normally remove unused categories in AnnData objects.

@Koncopd, would you agree with this? That this example should not error and the unused category should be removed?

```python
from itertools import cycle, islice
import pandas as pd
import scanpy as sc

pbmc = sc.datasets.pbmc3k_processed()
pbmc.obs["cat"] = pd.Categorical(islice(cycle("ab"), pbmc.n_obs), categories=list("abc"))

 sc.tl.rank_genes_groups(pbmc, groupby="cat")
```

```pytb
WARNING: Default of the method has been changed to 't-test' from 't-test_overestim_var'
---------------------------------------------------------------------------
ZeroDivisionError                         Traceback (most recent call last)
<ipython-input-7-d4efcd9bc643> in <module>
----> 1 sc.tl.rank_genes_groups(pbmc, groupby="cat")

~/github/scanpy/scanpy/tools/_rank_genes_groups.py in rank_genes_groups(adata, groupby, use_raw, groups, reference, n_genes, rankby_abs, pts, key_added, copy, method, corr_method, tie_correct, layer, **kwds)
    583     logg.debug(f'with sizes: {np.count_nonzero(test_obj.groups_masks, axis=1)}')
    584 
--> 585     test_obj.compute_statistics(
    586         method, corr_method, n_genes_user, rankby_abs, tie_correct, **kwds
    587     )

~/github/scanpy/scanpy/tools/_rank_genes_groups.py in compute_statistics(self, method, corr_method, n_genes_user, rankby_abs, tie_correct, **kwds)
    364         n_genes = self.X.shape[1]
    365 
--> 366         for group_index, scores, pvals in generate_test_results:
    367             group_name = str(self.groups_order[group_index])
    368 

~/github/scanpy/scanpy/tools/_rank_genes_groups.py in t_test(self, method)
    186         from scipy import stats
    187 
--> 188         self._basic_stats()
    189 
    190         for group_index, mask in enumerate(self.groups_masks):

~/github/scanpy/scanpy/tools/_rank_genes_groups.py in _basic_stats(self)
    171                 continue
    172 
--> 173             self.means[imask], self.vars[imask] = _get_mean_var(X_mask)
    174 
    175             if self.ireference is None:

~/github/scanpy/scanpy/preprocessing/_utils.py in _get_mean_var(X, axis)
      6 def _get_mean_var(X, *, axis=0):
      7     if sparse.issparse(X):
----> 8         mean, var = sparse_mean_variance_axis(X, axis=axis)
      9     else:
     10         mean = np.mean(X, axis=axis, dtype=np.float64)

~/github/scanpy/scanpy/preprocessing/_utils.py in sparse_mean_variance_axis(mtx, axis)
     40         )
     41     else:
---> 42         return sparse_mean_var_minor_axis(mtx.data, mtx.indices, *shape, np.float64)
     43 
     44 

ZeroDivisionError: division by zero
```

Reply:
Hm, i missed this discussion. This is another problem, isn't it? An empty group vs a group with only one element.
This happens because of [this](https://github.com/theislab/scanpy/blob/d20547460bc32882c707b21cdbe646ca8f8ce65f/scanpy/_utils.py#L488) i think.
However, shouldn't anndata care about this cases and not the functions which use `.obs`?

Reply:
If we could intercept things like adding categories, then we could do this in `anndata`, but then we'd have to create a wrapper class for dataframes.

I was wondering if we should just put a `sanitize_anndata` call at the top of `rank_genes_groups`.